examples/some-issuer/src/lib.rs

```diff
@@ -32,6 +33,69 @@ pub struct IssuerState {
     pub max_register_energy:   Energy,
     pub metadata_url:          Arc<Url>,
     pub credential_schema_url: Arc<str>,
+    pub rate_limiter:          Arc<RateLimiter>,
+}
+
+/// Remembers user ids of issued credentials up to a fixed capacity and how many
+/// times credentials have been issued for the same ids.
+///
+/// The implementation uses the reference count of [Arc](std::sync::Arc) to
+/// count the number of credentials that have been issued. So if the queue is at
+/// capacity, the last `Arc` is removed, decreasing the reference count and
+/// allowing one more credential for that id to be issued.
+pub struct RateLimiter {
+    /// MPMC FIFO queue that remembers the most recently
+    /// issued user ids.
+    queue:          crossbeam_deque::Injector<Arc<str>>,
+    /// Concurrent hashset that holds all elements of `queue`.
+    lookup:         DashSet<Arc<str>>,
+    queue_capacity: usize,
+    max_repeats:    usize,
+}
+
+impl RateLimiter {
+    pub fn new(queue_capacity: usize, max_repeats: usize) -> Self {
+        Self {
+            queue: crossbeam_deque::Injector::new(),
+            lookup: DashSet::new(),
+            queue_capacity,
+            max_repeats,
+        }
+    }
+
+    fn id_at_capacity(&self, id: String) -> bool {
+        let arc = Arc::<str>::from(id);
+        match self.lookup.get(&arc) {
+            // If there are self.max_repeats references in the queue and one in the hashset
+            Some(arc) if Arc::strong_count(&arc) > self.max_repeats => true,
+            _ => false,
+        }
+    }
+
+    fn insert(&self, id: String) {
+        let arc = Arc::<str>::from(id);
+        // Find the Arc for the id or use a new one, if the id does not exist
+        let arc = match self.lookup.get(&arc) {
+            Some(arc) => Arc::clone(&arc),
+            None => arc,
+        };
+
+        if self.queue.len() >= self.queue_capacity {
+            let mut removed = self.queue.steal();
+            while removed.is_retry() {
+                removed = self.queue.steal();
+            }
+            if let Some(removed) = removed.success() {
+                // Remove the entry from the map if there are only two references left
+                // (`removed` and the entry itsefl)
+                self.lookup
+                    .remove_if(&removed, |arc| Arc::strong_count(arc) == 2);
+            }
```

Comment:
I am not really sure that this is safe as such since we are not changing things atomically.

So suppose that you have the queue of size at most 3, and multiplicity 1

And you have the queue

`[A,B,C]`

And a request for `D` and `C` come in.

Then the request for `D` is handled first until line 93.

At this point request to insert C starts being handled. We now create a new `Arc` in line `80`. The condition in line `83` is no longer true. so we skip to line `96`.

We insert a new `Arc` into the `lookup` map. We then push it to the queue. Then the insertion of `D` resumes, we insert it into the `lookup` map. And push it to the queue. So the queue size is now one more than the maximum capacity.